### PR TITLE
Specs for invalid transaction filter.

### PR DESCRIPTION
**Pull request by smtmfft**
Checking if a TX is invalid by adding a few auxiliary columns to the TX table.
So far the balance, nonce and intrinsic gas checking is ready.
The access list gas cost checking is in fact a TODO because the specs does not enable the access list tx type (neither the circuit), so this part is more like a draft design for future use, and could be separated out if necessary.

Try addressing https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/872#discussion_r1022863932

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
Hi @Brechtpd, can you assign a reviewer for this PR?

**Comment by Brechtpd** [comment]
I'll do the review.

**Comment by ed255** [comment]
I think this is not considered in this PR but I'm wondering the following: for Taiko, have you consider the possibility to include a tx with invalid signature in the block?  Or do you consider the signature verification a pre-check to reject transactions before they are considered to be included in a block?

**Comment by Brechtpd** [comment]
It is indeed not part of this PR, but if needed I think we can add support for that in another PR if others are interested.

We will be handling blocks containing invalid transactions (except insufficient ETH/incorrect nonce transactions as they depend on the latest state) as completely invalid (a block with transactions like that won't have any state changes). In our system the proposer has full control over the included transactions so can make sure the transaction data/signature/etc... is valid, only the parts depending on the latest state (nonce, ETH balance) may be invalid because that's out of the proposer's control sometimes, so we want to be able to skip over individual transactions for that.

But we also still have to be able to prove blocks containing any kind of transaction data. But instead of having to handle all invalid cases in the circuit, we'll have a smart contract on L2 that will be usable to prove that the block contains invalid transactions, and so a proof for an invalid block is simply a normal proof with a single tx that calls this smart contract function with the block transaction data and shows a tx in the block is indeed invalid somehow (e.g. its signature is invalid, wrong RLP encoding etc...). This way the circuits remain simpler and we reuse the EVM itself with easier to write solidity code to handle these error cases (which normally will never be used because a proposer will always lose money by submitting blocks that contain these transactions that are always invalid).

File: specs/tables.md [hunk 1]
```diff
@@ -21,11 +21,15 @@ Proved by the tx circuit.
...
+| $TxID  | TxInvalid           | 0          | $value  |
+| $TxID  | AccessListGasCost   | 0          | $value  |
 | $TxID  | CallData            | $ByteIndex | $value  |
 | $TxID  | Pad                 | 0          | $value  |
 
-NOTE: `CallDataGasCost` and `TxSignHash` are values calculated by the verifier
-and used to reduce the circuit complexity.  They may be removed in the future.
+NOTE:
+- `CallDataGasCost` and `TxSignHash` are values calculated by the verifier and used to reduce the circuit complexity.  They may be removed in the future.
+- `TxInvalid` is a flag to tell the circuit which tx should be removed from tx processing. We check `balance`, `nonce`, and `intrinsic gas` within `begin_tx`, `end_tx`, `end_block` steps to make sure all tx being processed is valid, and all the others are invalid.
```
Comment (Brechtpd) [inline specs/tables.md:31]:
```suggestion
- `TxInvalid` is a flag to tell the circuit which tx is invalid and should not be executed. We check `balance`, `nonce`, and `intrinsic gas` within `begin_tx`, `end_tx`, and `end_block` steps to make sure all txs being processed are valid, and all others are invalid and not executed. Invalid transactions go from the `begin_tx` state directly to the `end_tx` state and do not have any side effects.
```

File: specs/tables.md [hunk 2]
```diff
@@ -21,11 +21,15 @@ Proved by the tx circuit.
...
+- `AccessListGasCost` is the accesslist gas cost of the tx, which equals to `sum([G_accesslistaddress + G_accessliststorage * len(TA[j]) for j in len(TA)])`. Ref (EIP 2930)
```
Comment (Brechtpd) [inline specs/tables.md:32]:
```suggestion
- `AccessListGasCost` is the `accessList` gas cost of the tx, which equals to `sum([G_accesslistaddress + G_accessliststorage * len(TA[j]) for j in len(TA)])` (EIP 2930).
```

File: src/zkevm_specs/evm/execution/begin_tx.py [hunk 3]
```diff
@@ -40,27 +41,38 @@ def begin_tx(instruction: Instruction):
...
-        tx_gas.expr()
-        - (GAS_COST_CREATION_TX if tx_is_create == 1 else GAS_COST_TX)
-        - tx_call_data_gas_cost.expr()
-    )
-    instruction.constrain_gas_left_not_underflow(gas_left)
+    tx_accesslist_gas = instruction.tx_context_lookup(tx_id, TxContextFieldTag.AccessListGasCost)
+    tx_intrinsic_gas = tx_calldata_gas_cost.expr() + tx_cost_gas + tx_accesslist_gas.expr()
+
+    # check instrinsic gas
+    gas_not_enough = int(tx_gas.expr()) < int(tx_intrinsic_gas)
```
Comment (Brechtpd) [inline src/zkevm_specs/evm/execution/begin_tx.py:56]:
```suggestion
    MAX_N_BYTES = 31 
    gas_not_enough, _ = instruction.compare(tx_gas, tx_intrinsic_gas, MAX_N_BYTES)
```

File: src/zkevm_specs/evm/execution/begin_tx.py [hunk 4]
```diff
@@ -40,27 +41,38 @@ def begin_tx(instruction: Instruction):
...
         tx_caller_address,
         tx_callee_address,
-        tx_value,
-        gas_fee,
+        tx_value if is_tx_invalid == 0 else RLC(0),
+        gas_fee if is_tx_invalid == 0 else RLC(0),
         reversion_info,
     )
+    sender_balance_prev = sender_balance_pair[1]
+    balance_not_enough = sender_balance_prev.int_value < tx_value.int_value + gas_fee.int_value
```
Comment (Brechtpd) [inline src/zkevm_specs/evm/execution/begin_tx.py:72]:
```suggestion
    balance_not_enough, _ = instruction.compare(
      instruction.rlc_to_fq(sender_balance_prev, MAX_N_BYTES),
      instruction.rlc_to_fq(tx_value, MAX_N_BYTES) + instruction.rlc_to_fq(gas_fee, MAX_N_BYTES),
      MAX_N_BYTES,
```

File: src/zkevm_specs/evm/execution/begin_tx.py [hunk 5]
```diff
@@ -28,10 +28,11 @@ def begin_tx(instruction: Instruction):
     instruction.constrain_not_zero(tx_caller_address)
 
     # Verify nonce
+    is_tx_invalid = instruction.tx_context_lookup(tx_id, TxContextFieldTag.TxInvalid)
     tx_nonce = instruction.tx_context_lookup(tx_id, TxContextFieldTag.Nonce)
     nonce, nonce_prev = instruction.account_write(tx_caller_address, AccountFieldTag.Nonce)
-    instruction.constrain_equal(tx_nonce, nonce_prev)
-    instruction.constrain_equal(nonce, nonce_prev.expr() + 1)
+    is_nonce_valid = instruction.is_zero(tx_nonce.expr() - nonce_prev.expr())
+    instruction.constrain_equal(nonce, nonce_prev.expr() + 1 - is_tx_invalid.expr())
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/begin_tx.py:35]:
```suggestion
    # bump the account nonce if the tx is valid
    instruction.constrain_equal(nonce, nonce_prev.expr() + 1 - is_tx_invalid.expr())
```

File: src/zkevm_specs/evm/execution/begin_tx.py [hunk 6]
```diff
@@ -40,27 +41,43 @@ def begin_tx(instruction: Instruction):
...
+    tx_calldata_gas_cost = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataGasCost)
+    tx_cost_gas = GAS_COST_CREATION_TX if tx_is_create == 1 else GAS_COST_TX
     # TODO: Handle gas cost of tx level access list (EIP 2930)
-    tx_call_data_gas_cost = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataGasCost)
-    gas_left = (
-        tx_gas.expr()
-        - (GAS_COST_CREATION_TX if tx_is_create == 1 else GAS_COST_TX)
-        - tx_call_data_gas_cost.expr()
-    )
-    instruction.constrain_gas_left_not_underflow(gas_left)
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/begin_tx.py:50]: Why do we no longer check gas_left underflow?

File: src/zkevm_specs/evm/execution/begin_tx.py [hunk 7]
```diff
@@ -40,27 +41,43 @@ def begin_tx(instruction: Instruction):
...
 
     # Verify transfer
-    instruction.transfer_with_gas_fee(
+    sender_balance_pair, _ = instruction.transfer_with_gas_fee(
         tx_caller_address,
         tx_callee_address,
-        tx_value,
-        gas_fee,
+        tx_value if is_tx_invalid == 0 else RLC(0),
+        gas_fee if is_tx_invalid == 0 else RLC(0),
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/begin_tx.py:69]:
```suggestion
        (1 - is_tx_invalid) * tx_value,
        (1 - is_tx_invalid) * gas_fee,
```

**Review by ChihChengLiang — COMMENTED** [review]
First pass

File: src/zkevm_specs/evm/execution/begin_tx.py (hunk 6 above)
Comment (smtmfft) [reply src/zkevm_specs/evm/execution/begin_tx.py:50]: There is another flag gas_not_enough to check underflow

Comment (ChihChengLiang) [reply src/zkevm_specs/evm/execution/begin_tx.py:50]: Oh, I see. For invalid transactions, the underflow is allowed.

**Review by ChihChengLiang — APPROVED** [review]
LGTM

File: src/zkevm_specs/evm/execution/begin_tx.py (hunk 7 above)
Comment (smtmfft) [reply src/zkevm_specs/evm/execution/begin_tx.py]: Need to revert this due to lint check failure. transfer_with_gas_fee asks for RLC input type.

**Comment by smtmfft** [comment]
Anyone know why I got this error??

==================================== ERRORS ====================================
__________________ ERROR collecting tests/evm/test_end_tx.py ___________________
tests/evm/test_end_tx.py::test_end_tx: in "parametrize" the number of names (6):
  ['tx', 'gas_left', 'refund', 'is_last_tx', 'current_cumulative_gas_used', 'success']
must be equal to the number of values (5):
  (<zkevm_specs.evm.typing.Transaction object at 0x7fe9028b8190>, 60000, 0, False, 0)
=========================== short test summary info ============================

There is no "success" parameter in test_end_tx.py. https://github.com/privacy-scaling-explorations/zkevm-specs/blob/2c7067605eb2dc5406bb50d0abe0962fd0feafec/tests/evm/test_end_tx.py#L82-L85

**Comment by CPerezz** [comment]
> zkevm_specs.evm.typing.Transaction object at 0x7fe9028b8190>,

That's really weird... The `parametrize` seems to match perfectly the Testing Data passed..
Maybe @han0110 or @ChihChengLiang can bring some light here

**Comment by ChihChengLiang** [comment]
Hey @smtmfft and @CPerezz, it looks like the test error is triggered when the PR is merged with the master branch. If you rebase the branch, the error can be reproduced.

**Comment by smtmfft** [comment]
> Hey @smtmfft and @CPerezz, it looks like the test error is triggered when the PR is merged with the master branch. If you rebase the branch, the error can be reproduced.

Thanks @ChihChengLiang , The pytest is fixed after rebase :)

**Comment by ChihChengLiang** [comment]
Hi @Brechtpd, I noticed you still need to approve the PR. Anything you'd like to add to the review?

**Review by Brechtpd — APPROVED** [review]
Nope, LGTM!

**Pull request merged**